### PR TITLE
added consumer group support to Reader

### PR DESCRIPTION
**Pull request by savaki**
I thought I would open a pull request early to get your initial feedback on the request.  This is a working version of consumer group support for the Reader.  I've manually verified interoperability with Sarama.  In addition, I've included an automated  compatibility test with sarama.

Almost all the changes are additive.  The main change was around (*Reader).start that now supports multiple readers and an initial offset.  That said, I tried to preserve the spirit of what I believe start was doing.

Would love to get your thoughts.

### CODE REVIEW

File: reader.go [hunk 1]
```diff
@@ -484,28 +1462,47 @@ func (r *Reader) readLag(ctx context.Context) {
...
+				maxWait:     r.config.MaxWait,
+				version:     r.version,
+				msgs:        r.msgs,
+				stats:       &r.stats,
+			}).run(ctx, offset)
+		}(ctx, partition, offset, &r.join)
+	}
+}
+
+func (r *Reader) CommitMessage(m Message) error {
```
Comment (achille-roussel) [inline reader.go]: Should we do this automatically as part of the `ReadMessage` method? I feel like otherwise a program has keep awareness of whether a Reader is part of a consumer group to know if it must call `CommitMessage` or not.

I feel like managing the commits automatically would better be inline with the current design where the offsets are already automatically managed when directly addressing partitions.

If this is a critical feature for one of your systems I think we should introduce a new method (say `PeekMessage`) which doesn't do automatic offset management and leaves it up to the program to explicitly commit messages.

File: reader.go [hunk 2]
```diff
@@ -824,3 +1819,25 @@ func (r *reader) withErrorLogger(do func(*log.Logger)) {
 		r.withLogger(do)
 	}
 }
+
+// extractTopics returns the unique list of topics represented by the set of
+// provided members
+func extractTopics(members []memberGroupMetadata) []string {
+	visited := map[string]struct{}{}
```
Comment (achille-roussel) [inline reader.go]: Style: keeping one variable declaration style within the local scope helps with readability, for example here you could rewrite this as
```go
var visited = map[string]struct{}{}
var topics []string
```

File: strategy.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+// strategy encapsulates the client side rebalancing logic
+type strategy interface {
+	// ProtocolName of strategy
+	ProtocolName() string
+
+	// ProtocolMetadata returns the metadata for the protocol
+	ProtocolMetadata(topics []string) (groupMetadata, error)
+
+	// DefineMemberships returns which members will be consuming
```
Comment (achille-roussel) [inline strategy.go]: This documentation seems incorrect.

File: strategy.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
+package kafka
+
+import "sort"
+
+// strategy encapsulates the client side rebalancing logic
+type strategy interface {
+	// ProtocolName of strategy
+	ProtocolName() string
```
Comment (achille-roussel) [inline strategy.go]: Any reason not to non-exported method names in this interface? (since the interface is non-exported itself)

File: syncgroup.go [hunk 5]
```diff
@@ -37,6 +37,10 @@ func (t groupAssignment) writeTo(w *bufio.Writer) {
 }
 
 func (t *groupAssignment) readFrom(r *bufio.Reader, size int) (remain int, err error) {
+	if size == 0 {
```
Comment (achille-roussel) [inline syncgroup.go]: Are there valid use cases where there will be no more bytes left to read even tho we expect to read a `groupAssignment` structure?

File: reader.go [hunk 6]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+			}
+		}
+	}
+
+	return offsetsByPartition, nil
+}
+
+func (r *Reader) subscribe(subs map[string][]int32) error {
+	// always clear prior subscriptions
+	r.unsubscribe()
```
Comment (achille-roussel) [inline reader.go]: It seems like `subscribe` is only called it one place, can we move this call out of this method so we keep a single responsibility per method?

File: reader.go [hunk 7]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+
+	if len(subs[r.config.Topic]) == 0 {
```
Comment (achille-roussel) [inline reader.go]: Would this be better expressed as
```go
if _, exists := subs[r.config.Topic]; !exists {
    return nil
}
```

Checking the length assuming to get the zero-value if it doesn't exist seems less straight forward.

File: reader.go [hunk 8]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+func (r *Reader) waitThrottleTime(throttleTimeMS int32) {
+	if throttleTimeMS == 0 {
+		return
+	}
+
+	select {
+	case <-r.stctx.Done():
+		return
+
+	case <-time.After(time.Duration(throttleTimeMS) * time.Millisecond):
```
Comment (achille-roussel) [inline reader.go]: Can we replace the use of `time.After` for a `timer.Timer`? That way we can stop it on return and we don't leak resources if `stctx` is canceled.

File: reader.go [hunk 9]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+				// offsets should only increase
+				continue
+			}
+
+			offsetsByPartition[m.Partition] = m.Offset
+		}
+	}
+}
+
+func (r *Reader) commitLoop(stop <-chan struct{}, done chan<- struct{}) {
```
Comment (achille-roussel) [inline reader.go]: Instead of a `stop` channel here we could use a context, that way it could inherit from `r.stctx` and we would get cancellation propagation if either the base context is canceled or only the one we created for the commit loop.

The same logic applies for the heartbeat loop.

File: reader.go [hunk 10]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+			return err
+		}
+	}
+
+	r.waitThrottleTime(resp.ThrottleTimeMS)
+
+	return nil
+}
+
+func (r *Reader) hbLoop(stop <-chan struct{}, done chan<- struct{}) {
```
Comment (achille-roussel) [inline reader.go]: How about naming this method `heartbeatLoop`? `hb*` is a bit hard to read.

File: reader.go [hunk 11]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+		assignments, err := r.rebalance()
+		if err != nil {
+			r.withErrorLogger(func(l *log.Logger) {
+				l.Printf("rebalance failed for consumer group, %v: %v", r.config.GroupID, err)
+			})
+			continue
+		}
+
+		// start the heartbeat
+		hbStop, hbDone := make(chan struct{}), make(chan struct{})
```
Comment (achille-roussel) [inline reader.go]: It seems like a `sync.WaitGroup` would be a better construct to wait for all sub-goroutines to complete instead of waiting on multiple "done" channels.

File: reader.go [hunk 12]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+
+		break
+	}
+
+	return nil
+}
+
+// newJoinGroupRequestV2 handles the logic of constructing a joinGroup
+// request
+func (r *Reader) newJoinGroupRequestV2() (joinGroupRequestV2, error) {
```
[the remark this answers is not in the document]
Comment (savaki) [reply reader.go]: Good to know.

File: reader.go (hunk 1 above)
Comment (savaki) [reply reader.go]: Perhaps we could go through some use cases.  

For us, it is critical that messages are processed before we commit the message off the Reader which requires an explicit commit.  Consider what would happen if one were performing db replication or event sourcing if a message that was read for reason failed to be processed.  Instead of INSERT > UPDATE > DELETE we could end up with just INSERT > UPDATE

That said, I can see some cases where keeping track of the offsets might not be as critical and a lossy, automatic commit would be acceptable.  Perhaps an AutoCommit flag on ReaderConfig?

File: strategy.go (hunk 4 above)
Comment (savaki) [reply strategy.go]: Two reasons, neither strong.  

The first is I have a personal pattern where I extend the Go style to apply to struct > struct communication.  Even if the type is lower case, I prefer the func to be uppercase to clearly indicate that it should be shared.

The second is that strategy is similar to Balancer, you can easily see others wanting to implement their own Strategies.  For example, a rack-aware consumer strategy.  In that case, uppercasing strategy is all that's required.

File: syncgroup.go (hunk 5 above)
Comment (savaki) [reply syncgroup.go]: I came across this case when testing for compatibility with bsm/sarama-cluster.  It appears in some cases, sarama-cluster can send a nil array entry.  Admittedly, I didn't look too closely at it.

File: reader.go (hunk 8 above)
Comment (savaki) [reply reader.go]: Sure, happy to change it. 

I figured it wouldn't be a problem since the loop is slow and resources are freed once time.After fires.  It's not so much a leak as holding resources longer than necessary.

File: reader.go (hunk 7 above)
Comment (savaki) [reply reader.go]: Happy to change it.  Just my idiosyncrasy.  Feels a little awkward to leave an edge case open that someone in the future could stumble upon.

File: reader.go (hunk 6 above)
Comment (savaki) [reply reader.go]: yep.  felt a little awkward to me as well, but I saw

```
r.cancel() // always cancel the previous reader
```
 
in start which seemed conceptually similar.

File: reader.go [hunk 13]
```diff
@@ -29,31 +66,847 @@ type Reader struct {
...
+			})
+			return nil, err
+
+			// all possible error codes returned by join group
+		//case GroupLoadInProgress:
+		//case GroupCoordinatorNotAvailable:
+		//case NotCoordinatorForGroup:
+		//case InconsistentGroupProtocol:
+		//case InvalidSessionTimeout:
+		//case GroupAuthorizationFailed:
```
[the remark this answers is not in the document]
Comment (savaki) [reply reader.go]: Makes sense.

File: reader.go [hunk 14]
```diff
@@ -29,31 +66,770 @@ type Reader struct {
...
+			ProtocolName:     strategy.ProtocolName(),
+			ProtocolMetadata: meta.bytes(),
+		})
+	}
+
+	return request, nil
+}
+
+// newMemberProtocolMetadata maps encoded member metadata ([]byte) into memberGroupMetadata
+func (r *Reader) newMemberProtocolMetadata(in []joinGroupResponseMemberV2) ([]memberGroupMetadata, error) {
```
Comment (achille-roussel) [inline reader.go]: Should we rename this “make” as well? It returns a value from a make call, seems like it would make sense.

File: reader.go [hunk 15]
```diff
@@ -29,31 +66,770 @@ type Reader struct {
...
+	}
+
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("joinGroup succeeded for response, %v.  generationID=%v, memberID=%v\n", r.config.GroupID, response.GenerationID, response.MemberID)
+	})
+
+	return assignments, nil
+}
+
+func (r *Reader) newSyncGroupRequestV1(memberAssignments memberGroupAssignments) syncGroupRequestV1 {
```
Comment (achille-roussel) [inline reader.go]: Same here, seems like this method should be named “make*”

File: reader.go [hunk 16]
```diff
@@ -297,6 +1178,10 @@ func (r *Reader) ReadMessage(ctx context.Context) (Message, error) {
 // The function returns a lag of zero when the reader's current offset is
 // negative.
 func (r *Reader) ReadLag(ctx context.Context) (lag int64, err error) {
+	if r.config.GroupID != "" {
```
Comment (achille-roussel) [inline reader.go]: This is tested plenty of times, how about adding a method to help make the code more readable (something like r.useConsumerGroup())

File: reader.go (hunk 1 above)
Comment (achille-roussel) [reply reader.go]: Do you have a concern about adding a new method that doesn’t do the auto-commit?

I feel like that would make it explicit whether CommitMessage has to be called or not. I would really prefer if we can keep the auto-commit behavior in ReadMessage. 

Also what do you think about changing the commit method to be called CommitMessages and accept an arbitrary number of messages as argument? That way the program acquired some ownership over the batching of commits (if we can make the commits synchronized I can see this being useful for some of our use cases at Segment, but we must get the API right to allow this).

File: strategy.go (hunk 4 above)
Comment (achille-roussel) [reply strategy.go]: Fair, I can see us exposing the strategy interface at some point :+1:

File: syncgroup.go (hunk 5 above)
Comment (achille-roussel) [reply syncgroup.go]: How about leaving a comment about it here? So anyone reading this code knows why we’re making the check.

File: reader.go (hunk 1 above)
Comment (achille-roussel) [reply reader.go]: PS: what do you think about passing a context as well to the CommitMeasages method? That way we also give control to the application on how long it may wait for the commit to be published?

Comment (savaki) [reply reader.go]: I think I'm just very used to thinking of read messages as being a singular activity and autocommit as a property of the connection rather than having ReadMessage vs PeekMessage/CommitMessage.  How would you feel about FetchMessage/CommitMessage for naming which would be more in sync with Kafka's naming.

Comment (savaki) [reply reader.go]: Here are the commit use cases that I'm thinking of.

1. fetch message, process message, commit message used by most of our use cases.  Since we require all services support at least once processing, async commits are fine since replays will be ignored.

2. fetch/process/commit for exactly once processing.  I'm not sure this is a thing with the api as we have it.  I believe Confluent created a new api to support these semantics, but admittedly I haven't read up on it.  Exactly once processing would really have to be at least once since there's no two phase commit.

3. fetch/process/batch-commit for micro-batch processing.

4. fetch/process/no-commit e.g. live statistics.  

4 doesn't require a commit, so that's easy.  1 & 2 only require a single message commit where 3 requires an arbitrarily sized batch.  

What about having two commit methods?  Sync/async could still be controlled by the CommitInterval field.

```
Commit(ctx context.Context, offsets Offsets) error
CommitMessage(m kafka.Message)
```

1 & 2 seems like they would almost always want to an async commit with a single message.  Having a ctx seems like it would be confusing.  

3 seems like it could benefit from a sync commit commit.  By separating the calls, we could also separate the shape of the data being sent.

For comparison, sarama-cluster uses the following:

```
func (c *Consumer) MarkOffset(msg *sarama.ConsumerMessage, metadata string)
func (c *Consumer) MarkOffsets(s *OffsetStash)
```

File: syncgroup.go (hunk 5 above)
Comment (savaki) [reply syncgroup.go]: Sure, that makes sense.

File: reader.go (hunk 1 above)
Comment (achille-roussel) [reply reader.go]: What I really want to avoid is creating a dependency between how the Reader is configured and how it is used, which is why I would prefer to have auto commits on ReadMessage. That way, no matter what the configuration is, the code does not need to be changed. Otherwise, the code using the Reader needs to be aware of its configuration to know whether it has to commit or not, or it has to maybe ignore the error, or we make the commit a no-op if it’s not needed? All those patterns seem less robust than splitting the APIs to have the program explicitly control how it consumes messages when it needs to.

FetchMessage sounds good to me (I actually like it better than Peek).

I feel like making the commit sync or async can be configured on the Reader, because in this case it doesn’t impact the way the program uses the code, it only makes a blocking operation non-blocking. This is actually the approach we took in the Writer to configure whether the writes are done synchronously it asynchronously. 

I’m not concerned about passing a context for cases where commits are asynchronous, very likely the program would use the same context that it passed to FetchMessage.

I’d like to avoid introducing a new Offset concept here, I liked your proposal of passing the message to commit, it makes it trivial to use the API, otherwise the program needs to extract the offsets etc...

It seems to me that a method with this signature would cover all use cases:
```go
CommitMessages(ctx context.Context, msg ...Message) error
```
(Plus it’s consistent with the model used with WriteMessages in the Writer)

Comment (savaki) [reply reader.go]: Works for me.

**Comment by savaki** [comment]
Sure, would be happy to.  There's additional tests I'd like to create on handling consumer groups under various edge cases, but I'd be good with merging as well.

**Comment by achille-roussel** [comment]
Cool, let's not block the merge then, feel free to send follow up PRs to add more tests 👍 

Thanks a ton for your contribution on this, it's massive 🙇
